### PR TITLE
BC-497 Alert when Provider Details API is unavailable during office hours

### PR DESCRIPTION
**Pull request by jackheslop96**
## BC-497

[Link to story](https://dsdmoj.atlassian.net/browse/BC-497)

- Downgraded provider API failure log level from warn to info, as alerting is now handled via Prometheus rather than log-based monitoring
- Added a Prometheus alert (ProviderDetailsApiUnavailable) that fires when the Provider Details API returns errors for 5+ minutes during office hours (Mon–Fri, 07:00–9:30 BST)

## Checklist

Before you ask people to review this PR:

- [x] Branch naming followed as per [LAA Ways Of Working](https://dsdmoj.atlassian.net/wiki/spaces/LP1/pages/5697536341/LAA+Ways+of+working#Core-Branches).
- [x] Tests should be passing: `./gradlew test`
- [x] Github should not be reporting conflicts; you should have recently run `git rebase main`.
- [x] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [x] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [x] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: src/main/java/uk/gov/justice/laa/amend/claim/service/ProviderService.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * Fetches provider firm name from Provider API
+     *
+     * @param officeAccountNumber the office account number
+     * @return firm name from API or office account number as fallback
+     */
+    public ProviderFirmOfficeDto getProviderFirm(String officeAccountNumber) {
+        LocalTime start = providerApiProperties.getStart().toLocalTime();
+        LocalTime end = providerApiProperties.getEnd().toLocalTime();
+        if (isInTimeRange(LocalTime.now(clock), start, end)) {
```
Comment (mboyd-moj) [inline src/main/java/uk/gov/justice/laa/amend/claim/service/ProviderService.java]: Can we have an info log if it's out of hours

File: src/main/java/uk/gov/justice/laa/amend/claim/utils/TimeUtils.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public final class TimeUtils {
+
+    public static boolean isInTimeRange(LocalTime now, LocalTime start, LocalTime end) {
+        if (start.equals(end)) {
+            return true;
+        }
+        if (start.isAfter(end)) {
```
Comment (mboyd-moj) [inline src/main/java/uk/gov/justice/laa/amend/claim/utils/TimeUtils.java]: Ooo nice thinking

File: src/main/java/uk/gov/justice/laa/amend/claim/utils/TimeUtils.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+            return now.isAfter(start) || now.isBefore(end);
+        }
+        return now.isAfter(start) && now.isBefore(end);
```
Comment (mboyd-moj) [inline src/main/java/uk/gov/justice/laa/amend/claim/utils/TimeUtils.java]: I think we probably want our comparisons to be inclusive. If it's 7.00am I'd say that's in the time range

File: src/test/java/uk/gov/justice/laa/amend/claim/utils/TimeUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    public void isInTimeRange_whenStartAndEndAreTheSame_returnsTrue() {
+        LocalTime now = LocalTime.of(12, 0);
+        LocalTime start = LocalTime.of(15, 0);
+        LocalTime end = LocalTime.of(15, 0);
+
+        Assertions.assertTrue(TimeUtils.isInTimeRange(now, start, end));
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {18, 19, 20, 21, 22, 23, 0, 1, 2, 3, 4, 5, 6, 7, 8})
```
Comment (mboyd-moj) [inline src/test/java/uk/gov/justice/laa/amend/claim/utils/TimeUtilsTest.java]: I think it'd be fine to just test the edge cases

File: src/main/java/uk/gov/justice/laa/amend/claim/config/ClockConfig.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import java.time.Clock;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class ClockConfig {
+
+    @Bean
+    Clock clock() {
+        return Clock.systemDefaultZone();
```
Comment (mboyd-moj) [inline src/main/java/uk/gov/justice/laa/amend/claim/config/ClockConfig.java]: Which timezone is this out of interest?

File: src/main/java/uk/gov/justice/laa/amend/claim/utils/TimeUtils.java (hunk 3 above)
Comment (jackheslop96) [reply src/main/java/uk/gov/justice/laa/amend/claim/utils/TimeUtils.java]: `now` will include seconds, nanoseconds. Highly unlikely we will ever get equality here. Also the `LocalTime` library doesn't contain `isAfterOrEqualTo`, `isBeforeOrEqualTo` etc. so it would be a lot of `!isBefore`, `!isAfter`

File: src/main/java/uk/gov/justice/laa/amend/claim/config/ClockConfig.java (hunk 5 above)
Comment (jackheslop96) [reply src/main/java/uk/gov/justice/laa/amend/claim/config/ClockConfig.java]: So technically (in the higher environments) this would be the timezone of the EC2 instance. We can more specifically set this as `Clock.systemUTC()`. I guess we should really align with how the provider details team are setting their OOH schedule.

Comment (mboyd-moj) [reply src/main/java/uk/gov/justice/laa/amend/claim/config/ClockConfig.java]: I'd like to keep ours as systemUTC and if we need to have a different timezone for provider team we pass that in config

File: src/main/java/uk/gov/justice/laa/amend/claim/utils/TimeUtils.java (hunk 3 above)
Comment (jackheslop96) [reply src/main/java/uk/gov/justice/laa/amend/claim/utils/TimeUtils.java]: So it would be
```
public static boolean isInTimeRange(LocalTime now, LocalTime start, LocalTime end) {
        if (start.equals(end)) {
            return true;
        }
        if (start.isAfter(end)) {
            return !now.isBefore(start) || !now.isAfter(end);
        }
        return !now.isBefore(start) && !now.isAfter(end);
    }
```
or
```
public static boolean isInTimeRange(LocalTime now, LocalTime start, LocalTime end) {
        if (start.equals(end)) {
            return true;
        }
        if (start.isAfter(end)) {
            return !(now.isBefore(start) && now.isAfter(end));
        }
        return !(now.isBefore(start) || now.isAfter(end));
    }
```
personally I think that affects readability but happy to change it if we see fit

Comment (mboyd-moj) [reply src/main/java/uk/gov/justice/laa/amend/claim/utils/TimeUtils.java]: I think as a utility method this would be better. I know at the moment we're passing in system time and it's almost zero chance we're on the boundary but we could use this for checking something else in the future and forget that we're not inclusive

File: src/main/java/uk/gov/justice/laa/amend/claim/service/ProviderService.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    private boolean isAvailable() {
+        boolean available;
+        try {
+            HealthDto response = providerApiClient.ping().block();
+            available = response != null && response.getStatus().equals(Status.UP);
+        } catch (Exception e) {
+            available = false;
+        }
+        if (!available) {
+            log.info("Provider details API is unavailable");
```
Comment (mboyd-moj) [inline src/main/java/uk/gov/justice/laa/amend/claim/service/ProviderService.java]: Can we configure an alert for this occurring during office hours? That way we'll at least notice if their API goes down

File: src/main/java/uk/gov/justice/laa/amend/claim/models/HealthDto.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package uk.gov.justice.laa.amend.claim.models;
+
+import lombok.Data;
+import org.springframework.boot.health.contributor.Status;
+
+@Data
+public class HealthDto {
```
Comment (mboyd-moj) [inline src/main/java/uk/gov/justice/laa/amend/claim/models/HealthDto.java]: Would a record or `@Value` suffice?

Comment (jackheslop96) [reply src/main/java/uk/gov/justice/laa/amend/claim/models/HealthDto.java]: Yes good shout

File: src/main/java/uk/gov/justice/laa/amend/claim/service/ProviderService.java (hunk 6 above)
Comment (jackheslop96) [reply src/main/java/uk/gov/justice/laa/amend/claim/service/ProviderService.java]: Cool idea yes let me have a look in Sentry to see what can be done

Comment (jackheslop96) [reply src/main/java/uk/gov/justice/laa/amend/claim/service/ProviderService.java]: Discussions had around handling this in Sentry or Pingdom

File: src/main/java/uk/gov/justice/laa/amend/claim/service/ProviderService.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    private final ProviderApiClient providerApiClient;
+
+    /**
+     * Fetches provider firm office from Provider API
+     *
+     * @param officeAccountNumber the office account number
+     * @return firm office from API or null as fallback
+     */
+    public ProviderFirmOfficeDto getProviderFirm(String officeAccountNumber) {
+        if (isAvailable()) {
```
Comment (ahmadsjustice) [inline src/main/java/uk/gov/justice/laa/amend/claim/service/ProviderService.java]: Is the isAvailable() check adding any value?
Happy path (API is up): We make 2 calls instead of 1 - isAvailable() + the actual provider call.   
Out of hours (API is down): We fail on isAvailable() instead of failing on the actual provider call - same delay either way.
Without isAvailable(), the outcome is identical in both cases - we just skip the extra call on the happy path.

Comment (jackheslop96) [reply src/main/java/uk/gov/justice/laa/amend/claim/service/ProviderService.java]: @ahmadsjustice the key is that we don't want to be creating unnecessary noise with warning logs during out of hours that the lookup failed. But we do want to be alerted in-hours.

Sentry alerts are currently created for any warning log in a given window (think it's 5 minutes). The more alerts we have of this nature, the more likely it is that a genuine alert could be missed.

File: .helm/laa-amend-a-claim/templates/prometheus-rules.yaml [hunk 9]
```diff
@@ -84,4 +84,20 @@ spec:
...
+  - name: application-rules
+    rules:
+    - alert: ProviderDetailsApiUnavailable
+      annotations:
+        message: Provider Details API is returning errors during office hours.
+        dashboard_url: {{ .Values.alerting.grafanaDashboard }}
+      expr: >-
+        (
+          increase(http_client_requests_seconds_count{namespace="{{ .Release.Namespace }}", uri="/api/v1/provider-offices/{officeCode}", outcome!="SUCCESS"}[5m]) > 0
+        )
```
[the remark this answers is not in the document]
Comment (ahmadsjustice) [reply .helm/laa-amend-a-claim/templates/prometheus-rules.yaml:97]: interested in knowing each error - message now accurately describes the behaviour.

File: .helm/laa-amend-a-claim/templates/prometheus-rules.yaml [hunk 10]
```diff
@@ -84,4 +84,20 @@ spec:
...
+        and on() (hour() >= 7 and hour() <= 17)
+        and on() (day_of_week() >= 1 and day_of_week() <= 5)
+      for: 0m
+      labels:
```
[the remark this answers is not in the document]
Comment (ahmadsjustice) [reply .helm/laa-amend-a-claim/templates/prometheus-rules.yaml:101]: The intent is for this alert to be strict. We want to catch every error, including transient blips. The provider API being unavailable directly impacts the user experience

File: .helm/laa-amend-a-claim/templates/prometheus-rules.yaml [hunk 11]
```diff
@@ -84,4 +84,20 @@ spec:
...
+    rules:
+    - alert: ProviderDetailsApiUnavailable
+      annotations:
+        message: Provider Details API has returned errors in the last 5 minutes during office hours.
+        dashboard_url: {{ .Values.alerting.grafanaDashboard }}
+      expr: >-
+        (
+          increase(http_client_requests_seconds_count{namespace="{{ .Release.Namespace }}", uri="/api/v1/provider-offices/{officeCode}", outcome!="SUCCESS"}[5m]) > 0
+        )
+        and on() (hour() >= 7 and hour() <= 16)
```
Comment (BenMillar-MOJ) [inline .helm/laa-amend-a-claim/templates/prometheus-rules.yaml]: PDAs opening hours are 07:00  - 21:30

**Review by BenMillar-MOJ — CHANGES_REQUESTED** [review]

**Review by BenMillar-MOJ — APPROVED** [review]
